parcagpu/parcagpu.go

```diff
@@ -0,0 +1,146 @@
+package parcagpu // import "go.opentelemetry.io/ebpf-profiler/parcagpu"
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+	"unsafe"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/perf"
+	log "github.com/sirupsen/logrus"
+
+	"go.opentelemetry.io/ebpf-profiler/host"
+	"go.opentelemetry.io/ebpf-profiler/interpreter/gpu"
+	"go.opentelemetry.io/ebpf-profiler/support"
+)
+
+type gpuTraceFixer struct {
+	mu                  sync.Mutex
+	timesAwaitingTraces map[uint32]float32
+	tracesAwaitingTimes map[uint32]*host.Trace
+}
+
+func (p *gpuTraceFixer) addTrace(trace *host.Trace) *host.Trace {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	id := trace.ParcaGPUTraceID
+	millis, ok := p.timesAwaitingTraces[id]
+	if ok {
+		delete(p.timesAwaitingTraces, id)
+		trace.OffTime = int64(millis * 1000000.0)
+		return trace
+	}
+	p.tracesAwaitingTimes[id] = trace
+	return nil
+}
+
+func (p *gpuTraceFixer) addTime(id uint32, millis float32) *host.Trace {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	trace, ok := p.tracesAwaitingTimes[id]
+	if ok {
+		delete(p.tracesAwaitingTimes, id)
+		trace.OffTime = int64(millis * 1000000.0)
+		return trace
+	}
+	p.timesAwaitingTraces[id] = millis
+	return nil
+}
+
+// uprobes aren't perfect and we may miss matching timing to trace at attach boundary
```

Comment:
"uprobes aren't perfect" -- is this documented somewhere by Linux? Under what conditions can we expect them to fail to fire?